File: performance-comparison/benchmarks/haskell-project/src/Factorial.hs
-- factorial -- tail-recursive accumulator mod 1e9+7 at every step.
{-# LANGUAGE BangPatterns #-}
import System.Environment (getArgs)
modulus :: Int
modulus = 1000000007
fact :: Int -> Int -> Int
fact !n !acc | n <= 1 = acc `mod` modulus
             | otherwise = fact (n - 1) ((acc * n) `mod` modulus)
main :: IO ()
main = do
  args <- getArgs
  let n = case args of (a:_) -> read a; _ -> 1000 :: Int
  print (fact n 1)

File: performance-comparison/benchmarks/rust-workspace/src/bin/factorial.rs
// factorial -- tail-recursive-shape factorial mod 1e9+7, expressed as a
// loop (Rust gives no TCO guarantee for the recursive form).
use std::env;

const MODULUS: i64 = 1_000_000_007;

fn fact(n: i64) -> i64 {
    let mut acc: i64 = 1;
    let mut k = n;
    while k > 1 {
        acc = (acc * k) % MODULUS;
        k -= 1;
    }
    acc % MODULUS
}

fn main() {
    let n: i64 = env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(1000);
    println!("{}", fact(n));
}
